uefi-test-runner/src/proto/load.rs

```diff
@@ -0,0 +1,128 @@
+use alloc::boxed::Box;
+use alloc::string::{String, ToString};
+use alloc::vec::Vec;
+use core::ffi::c_void;
+use core::pin::Pin;
+use core::ptr;
+use core::ptr::addr_of;
+use uefi::prelude::BootServices;
+use uefi::proto::device_path::build::DevicePathBuilder;
+use uefi::proto::media::load_file::{LoadFile, LoadFile2};
+use uefi::{Guid, Handle};
+use uefi_raw::protocol::device_path::DevicePathProtocol;
+use uefi_raw::protocol::media::{LoadFile2Protocol, LoadFileProtocol};
+use uefi_raw::Status;
+
+unsafe extern "efiapi" fn raw_load_file(
+    this: *mut LoadFile2Protocol,
+    _file_path: *const DevicePathProtocol,
+    _boot_policy: bool,
+    buffer_size: *mut usize,
+    buffer: *mut c_void,
+) -> Status {
+    log::debug!("hello");
+    let this = this.cast::<CustomLoadFile2Protocol>().as_ref().unwrap();
+    this.load_file(buffer_size, buffer.cast())
+}
+
+#[repr(C)]
+struct CustomLoadFile2Protocol {
+    inner: LoadFile2Protocol,
+    file_data: Vec<u8>,
+}
+
+impl CustomLoadFile2Protocol {
+    fn new(file_data: Vec<u8>) -> Pin<Box<Self>> {
+        let inner = Self {
+            inner: LoadFile2Protocol {
+                load_file: raw_load_file,
+            },
+            file_data,
+        };
+        Box::pin(inner)
+    }
+
+    fn load_file(&self, buf_len: *mut usize, buf: *mut c_void) -> Status {
+        log::debug!("hello");
+        if buf.is_null() || unsafe { *buf_len } < self.file_data.len() {
+            log::debug!("hello");
+            unsafe { *buf_len = self.file_data.len() };
+            Status::BUFFER_TOO_SMALL
+        } else {
+            unsafe {
+                ptr::copy_nonoverlapping(self.file_data.as_ptr(), buf.cast(), self.file_data.len());
+            }
+            Status::SUCCESS
+        }
+    }
+}
+
+unsafe fn install_protocol(
+    bt: &BootServices,
+    handle: Handle,
+    guid: Guid,
+    protocol: &mut CustomLoadFile2Protocol,
+) {
+    bt.install_protocol_interface(Some(handle), &guid, addr_of!(*protocol).cast())
+        .unwrap();
+}
+
+unsafe fn uninstall_protocol(
+    bt: &BootServices,
+    handle: Handle,
+    guid: Guid,
+    protocol: &mut CustomLoadFile2Protocol,
+) {
+    bt.uninstall_protocol_interface(handle, &guid, addr_of!(*protocol).cast())
+        .unwrap();
+}
+
+/// This tests the LoadFile and LoadFile2 protocols. As this protocol is not
+/// implemented in OVMF for the default handle, we implement it manually using
```

Comment:
What is the "default handle"?

Reply:
`bt.image_handle()`. Any idea for a better phrasing?

Reply:
How about "not implemented in OVMF for any handle" -- I don't think we have any way to get a handle from OVMF with these protocols installed. (If I have that wrong, then we could just use that handle rather than the manual implementation here.)